File: stack/ReverseAString.cpp
// Method 1 
// push all the element to the other stack and pop then and push them back in the
// original stack
#include<bits/stdc++.h>
using namespace std;
class Stack{
    public:
    int top;
    unsigned capacity;
    char * array;
};

// function to create a stack of given capacity
Stack *createStack(unsigned capacity){
    Stack *stack= new Stack();
    stack->capacity = capacity;
    stack->top = -1;
    stack->array = new char[(stack->capacity * sizeof(char))];
    return stack;
}

// check if stack is full top= last index
int isFull(Stack * stack){
    return stack->top == stack->capacity-1;
}
// check if stack is empty top= -1
int isEmpty(Stack * stack){
    return stack->top == -1;
}

// function to add an item to stack
void push(Stack *stack,char item){
    if(isFull(stack))
        return;
    stack->array[++stack->top] = item;
}
// remove an item from stack
char pop(Stack *stack){
    if(isEmpty(stack))
        return -1;
    return stack->array[stack->top --];
}
void reverseI(char str[]){
    int n = strlen(str);
    Stack * stack = createStack(n);
    // Push all characters of string to stack
    int i;
    for(i=0;i<n;i++)
        push(stack,str[i]);
    for(i=0;i<n;i++)
        str[i]= pop(stack);
}
// Complexities
// O(n) time complexity
// O(n) Auxiliary space



// Method 2
// without auxiliary space
void swap(char *a,char *b){
    char temp=*a;
    *a=*b;
    *b=temp;
}
void reverseII(char str[]){
    int n=strlen(str),i;
    for(i=0;i<n/2;i++)
        swap(&str[i],&str[n-1-i]);
}
int main(){
    char str[]="Hello World";
    cout<<"Reversed String Method 1: ";
    reverseI(str);
    cout<<str;
    cout<<endl<<endl;
    cout<<"Reversed String Method 2: ";
    reverseII(str);
    cout<<str;
    cout<<endl<<endl;

}

File: StacksAndQueues/ReverseAString.java
// Reverse a String using Stack
public class ReverseAString {
    static class CharStack{
        int size;
        int top;
        char[] arr;
        CharStack(int n){
            size = n;
            top = -1;
            arr = new char[n];
        }
        boolean isEmpty(){
            return top<0;
        }
        boolean push(char x){
            if(top >=size){
                System.out.println("Stack Overflow");
                return false;
            }
            arr[++top] =x;
            return true;
        }
        char pop(){
            if(top<0){
                System.out.println("Stack Underflow");
                return 0;
            }
            char x = arr[top--];
            return x;
        }
    }
    static void reverse(StringBuffer str){
        int n = str.length();
        CharStack chst = new CharStack(n);
        int i;
        for(i=0;i<n;i++)
            chst.push(str.charAt(i));
        for(i = 0;i<n;i++){
            str.setCharAt(i, chst.pop());
        }
    }
    public static void main(String[] args) {
        StringBuffer s= new StringBuffer("Hello World");
        reverse(s);
        System.out.println("Reversed string is " + s);

    }
}
